grammar.js

```diff
@@ -11,22 +19,191 @@ module.exports = grammar({
+      // So we have to express variations that require at least one of these.
+      return choice(
+        // decimal required, everything else optional
+        /[+\-]?[0-9][0-9_]*\.[0-9_]+(e[+\-]?[0-9]+)?(f32|f64)?/,
+        // exponent required, everything else optional
+        /[+\-]?[0-9][0-9_]*(\.[0-9_]+)?e[+\-]?[0-9]+(f32|f64)?/,
+        // suffix required, everything else optional
+        /[+\-]?[0-9][0-9_]*(\.[0-9_]+)?(e[+\-]?[0-9]+)?(f32|f64)/
+      );
+    },
+
+    /** 
+     * @see {@link https://crystal-lang.org/reference/syntax_and_semantics/literals/integers.html}
+     */
+    integer: $ => choice(
+      // Binary numbers
+      /[+\-]?0b[01]+(i8|i16|i32|i64|u8|u16|u32|u64)?/,
+      // Octal numbers
+      /[+\-]?0o[0-7]+(i8|i16|i32|i64|u8|u16|u32|u64)?/,
+      // Hexadecimal numbers
+      /[+\-]?0x[0-9a-fA-F]+(i8|i16|i32|i64|u8|u16|u32|u64)?/, 
+      // Decimal numbers
+      /[+\-]?[1-9][0-9_]*(i8|i16|i32|i64|u8|u16|u32|u64)?/
     ),
 
     symbol: $ => seq(':', choice(
-      //        A-Z        _      a-z                  0-9      A-Z       _        a-z
-      /[^\x00-\x40\x5B-\x5E\x60-\x60\x7B-\x9F][^\x00-\x2F\x3A-\x40\x5B-\x5E\x60-\x60\x7B-\x9F]*[=!\?]?/,
+      identifierRegex,
       $._operator,
       seq('"', repeat(choice(/[^"]/, '\"')), '"'),
     )),
 
+    /**
+	   * @see {@link https://crystal-lang.org/reference/syntax_and_semantics/literals/char.html}
+	   */
+    char: $ => {
+      const charsThatCanBeEscaped = [
+        "'",
+        "\\", // this is supposed to represent one backslash, but we have to escape it here
+        "a",
+        "b",
+        "e",
+        "f",
+        "n",
+        "r",
+        "t",
+        "v"
+      ];
+      const escapedLiteralChars = charsThatCanBeEscaped.map(c => "'\\" + c + "'");
+      // For stuff like '\u0042' and '\u{04AF}'
+      const unicodeEscape = /'\\u{?[0-9A-Fa-f]+}?'/;
+      return choice(
+        "''", // empty char
+        ...escapedLiteralChars,
+        unicodeEscape,
+        /'[^']'/ // any single literal char
+      );
+    },
+
+    // TODO: string interpolation
     string: $ => seq('"', repeat(choice(/[^"]/, '\"')), '"'),
 
-    _operator: $ => choice( "+", "-", "*", "/", "%", "&", "|", "^", "**", ">>", "<<", "==", "!=", "<", "<=", ">", ">=", "<=>", "===", "[]", "[]?", "[]=", "!", "~", "!~", "=~",),
+    /**
+	   * @see {@link https://crystal-lang.org/reference/syntax_and_semantics/literals/array.html}
+	   */
+    array: $ => seq(
+      '[',
+      commaSep1($._expression),
+      ']'
+    ),
+
+    /**
+	   * @see {@link https://crystal-lang.org/reference/syntax_and_semantics/literals/hash.html}
+	   */
+    hash: $ => {
+      const keyToValue = seq(
+        field('key', $._expression), 
+        '=>', 
+        field('value', $._expression)
+      );
+      const nonEmptyHash = seq(
+        // TODO: handle the MyHash{"one" => 1, "two" => 2} case
```

Comment:
This "TODO" is here because there's not currently support in this grammar (yet!) for matching "type" expressions. I'm hoping to add that in a future PR and come back around to this, and make this addition:

```suggestion
       optional($.class_name), // or whatever the "term" for a class/type identifier gets called
```